cylc/uiserver/workflows_mgr.py

```diff
@@ -42,69 +43,84 @@
     """Workflow request command."""
-    if context is None:
-        context = command
+    if req_context is None:
+        req_context = command
     try:
         result = await client.async_request(command, args, timeout)
-        return (context, result)
+        return (req_context, result)
     except ClientTimeout as exc:
         logger.exception(exc)
-        return (context, MSG_TIMEOUT)
+        return (req_context, MSG_TIMEOUT)
     except ClientError as exc:
         logger.exception(exc)
-        return (context, None)
+        return (req_context, None)
 
 
-async def est_workflow(reg, host, port, timeout=None):
+async def est_workflow(reg, host, port, pub_port, context=None, timeout=None):
     """Establish communication with workflow, instantiating REQ client."""
     if is_remote_host(host):
         try:
             host = get_host_ip_by_name(host)  # IP reduces DNS traffic
         except socket.error as exc:
             if flags.debug:
                 raise
-            logger.error("ERROR: %s: %s\n" % (exc, host))
-            return (reg, host, port, None)
+            logger.error("ERROR: %s: %s\n", exc, host)
+            return (reg, host, port, pub_port, None)
 
     # NOTE: Connect to the suite by host:port. This way the
     #       SuiteRuntimeClient will not attempt to check the contact file
     #       which would be unnecessary as we have already done so.
     # NOTE: This part of the scan *is* IO blocking.
-    client = SuiteRuntimeClient(reg, host=host, port=port, timeout=timeout)
-    context, result = await workflow_request(client, 'identify')
-    return (reg, host, port, client, result)
+    client = SuiteRuntimeClient(reg, host=host, port=port,
+                                context=context, timeout=timeout)
+    _, result = await workflow_request(client, 'identify')
+    return (reg, host, port, pub_port, client, result)
 
 
-class WorkflowsManager(object):
+class WorkflowsManager:
+    """Discover and Manage workflows."""
 
-    def __init__(self):
+    def __init__(self, uiserver, context=None):
+        self.uiserver = uiserver
+        if context is None:
+            self.context = zmq.asyncio.Context()
+        else:
+            self.context = context
         self.workflows = {}
+        self.stopping = set()
 
     def spawn_workflow(self):
+        """Start/spawn a workflow."""
         # TODO - Spawn workflows
         pass
 
     async def gather_workflows(self):
+        """Scan, establish, and discard workflows."""
         scanflows = {}
         cre_owner, cre_name = re_compile_filters(None, ['.*'])
         scan_args = (
-            (reg, host, port, CLIENT_TIMEOUT)
-            for reg, host, port in
-            get_scan_items_from_fs(cre_owner, cre_name))
+            (reg, host, port, pub_port, self.context, CLIENT_TIMEOUT)
+            for reg, host, port, pub_port in
+            get_scan_items_from_fs(cre_owner, cre_name)
```

Comment:
Suites launched pre https://github.com/cylc/cylc-flow/pull/3389 will generate quite a lot of traceback somewhere around here. Might be worth suppressing it, the contact file has an `API` field which we can use to filter out incompatible suites.

```
2019-11-28 11:34:22,270 tornado.application ERROR    Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fb5a252b400>>, <Task finished coro=<WorkflowsManager.gather_workflows() done, defined at ~/cylc-uiserver/cylc/uiserver/workflows_mgr.py:99> exception=KeyError('CYLC_SUITE_PUBLISH_PORT')>)
Traceback (most recent call last):
  File "/home/h06/osanders/miniconda3/envs/cylc8/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
    ret = callback()
  File "/home/h06/osanders/miniconda3/envs/cylc8/lib/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
    future.result()
  File "~/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 112, in gather_workflows
    for arg in scan_args:
  File "~/cylc-uiserver/cylc/uiserver/workflows_mgr.py", line 104, in <genexpr>
    (reg, host, port, pub_port, self.context, CLIENT_TIMEOUT)
  File "~/cylc/cylc/flow/network/scan.py", line 280, in get_scan_items_from_fs
    contact_data[ContactFileFields.PUBLISH_PORT]
KeyError: 'CYLC_SUITE_PUBLISH_PORT'
```

Reply:
True.. I might have to increment the `cylc-flow` API version in the publish PR.

Reply:
This is fixed, however, in order to get the API version without connecting a client I added `api` to the result of `get_scan_items_from_fs`. So there will still be errors until this is in, but as long as scan items remain the same thereafter, then the UI Server will only talk to those workflows with the same version (and should (optionally) restart workflows to version, otherwise).

Reply:
Nice, thanks.